File: MyProject/Source/MyProject/Private/BasicFrame/Libs/Auxiliary/AuxLoader/AuxLoaderBase.cpp
#include "MyProject.h"
#include "AuxLoaderBase.h"
#include "ResLoadState.h"
#include "ResEventDispatch.h"

namespace MyNS
{
	AuxLoaderBase::AuxLoaderBase(std::string path)
	{
		this->mInitPath = path;
		this->mResLoadState = new ResLoadState();

		this->reset();
	}

	void AuxLoaderBase::reset()
	{
		this->mResLoadState->reset();
		this->mPrePath = "";
		this->mPath = "";
		this->mIsInvalid = true;
	}

	void AuxLoaderBase::dispose()
	{
		this->unload();
	}

	bool AuxLoaderBase::hasSuccessLoaded()
	{
		return this->mResLoadState->hasSuccessLoaded();
	}

	bool AuxLoaderBase::hasFailed()
	{
		return this->mResLoadState->hasFailed();
	}

	// 加载成功或者加载失败
	bool AuxLoaderBase::hasLoadEnd()
	{
		return (this->hasSuccessLoaded() || this->hasFailed());
	}

	// 是否需要卸载资源
	bool AuxLoaderBase::needUnload(std::string path)
	{
		return this->mPath != path && !UtilStr::IsNullOrEmpty(path) && !UtilStr::IsNullOrEmpty(this->mPath);
	}

	void AuxLoaderBase::setPath(std::string path)
	{
		this->mPrePath = mPath;
		this->mPath = path;

		if (this->mPrePath != this->mPath && !UtilStr::IsNullOrEmpty(this->mPath))
		{
			this->mIsInvalid = true;
		}
		else
		{
			this->mIsInvalid = false;
		}
	}

	void AuxLoaderBase::updatePath(std::string path)
	{
		if (this->needUnload(path))
		{
			this->unload();
		}

		this->setPath(path);
	}

	bool AuxLoaderBase::isInvalid()
	{
		return this->mIsInvalid;
	}

	std::string AuxLoaderBase::getLogicPath()
	{
		return this->mPath;
	}

	void AuxLoaderBase::addEventHandle(EventDispatchDelegate evtHandle)
	{
		if (nullptr != evtHandle)
		{
			if (nullptr == this->mEvtHandle)
			{
				this->mEvtHandle = new ResEventDispatch();
			}
			this->mEvtHandle->addEventHandle(evtHandle);
		}
	}

	void AuxLoaderBase::syncLoad(std::string path, EventDispatchDelegate evtHandle)
	{
		this->mResLoadState->setLoading();

		this->updatePath(path);

		this->addEventHandle(evtHandle);
	}

	void AuxLoaderBase::asyncLoad(std::string path, EventDispatchDelegate evtHandle)
	{
		this->mResLoadState->setLoading();

		this->updatePath(path);

		this->addEventHandle(evtHandle);
	}

	void AuxLoaderBase::download(std::string origPath, EventDispatchDelegate  dispObj, long fileLen, bool isWriteFile, int downloadType)
	{
		this->mResLoadState->setLoading();

		this->updatePath(origPath);

		this->addEventHandle(dispObj);
	}

	void AuxLoaderBase::unload()
	{
		if (this->mEvtHandle != nullptr)
		{
			this->mEvtHandle->clearEventHandle();
			this->mEvtHandle = nullptr;
		}

		this->reset();
	}
}

File: Project/MyProject/MyProject/Content/MyAsset/Lua/MyLua/Libs/Auxiliary/AuxLoader/AuxLoaderBase.lua
MLoader("MyLua.Libs.Core.GlobalNS");
MLoader("MyLua.Libs.Core.Class");
MLoader("MyLua.Libs.Core.GObject");

local M = GlobalNS.Class(GlobalNS.GObject);
M.clsName = "AuxLoaderBase";
GlobalNS[M.clsName] = M;

function M:ctor()
	self:reset();
end

function M:dtor()
    
end

function M:dispose()
	self:unload();
end

function M:reset()
	self.mPrePath = "";
	self.mPath = "";
	self.mIsInvalid = true;
	self.mEvtHandle = nil;
	self.mProgressEventDispatch = nil;
end

function M:unload()
	if (self.mEvtHandle ~= nil) then
		self.mEvtHandle:clearEventHandle();
		self.mEvtHandle = nil;
	end

	if (self.mProgressEventDispatch ~= nil) then
		self.mProgressEventDispatch:clearEventHandle();
		self.mProgressEventDispatch = nil;
	end

	self:reset();
end

function M:hasSuccessLoaded()
	return self:getNativeLoader():hasSuccessLoaded();
end

function M:hasFailed()
	return self:getNativeLoader():hasFailed();
end

function M:hasLoadEnd()
	return self:getNativeLoader():hasLoadEnd();
end

function M:getLogicPath()
	return self:getNativeLoader():getLogicPath();
end

function M:getNativeLoader()
	return nil;
end

-- 是否需要卸载资源
function M:needUnload(path)
	return self.mPath ~= path and not GlobalNS.UtilStr.IsNullOrEmpty(path) and not GlobalNS.UtilStr.IsNullOrEmpty(self.mPath);
end

function M:setPath(path)
	self.mPrePath = self.mPath;
	self.mPath = path;

	if(self.mPrePath ~= self.mPath and not GlobalNS.UtilStr.IsNullOrEmpty(self.mPath)) then
		self.mIsInvalid = true;
	else
		self.mIsInvalid = false;
	end
end

function M:updatePath(path)
	if (self:needUnload(path)) then
		self:unload();
	end

	self:setPath(path);
end

function M:isInvalid()
	return self.mIsInvalid;
end

-- 添加事件处理器
function M:addEventHandle(pThis, evtHandle, progressHandle)
	if (nil ~= evtHandle) then
		if (nil == self.mEvtHandle) then
			self.mEvtHandle = GlobalNS.new(GlobalNS.ResEventDispatch);
		end

		self.mEvtHandle:addEventHandle(pThis, evtHandle, 0);
	end

	if (nil ~= progressHandle) then
		if (nil == self.mProgressEventDispatch) then
			self.mProgressEventDispatch = GlobalNS.new(GlobalNS.AddOnceEventDispatch);
		end

		self.mProgressEventDispatch:addEventHandle(pThis, progressHandle, 0);
	end
end

function M:syncLoad(path, pThis, evtHandle, progressHandle)
	self:updatePath(path);
	self:addEventHandle(pThis, evtHandle, progressHandle);
end

function M:asyncLoad(path, pThis, evtHandle, progressHandle)
	self:updatePath(path);
	self:addEventHandle(pThis, evtHandle, progressHandle);
end

function M:download(origPath, pThis, evtHandle, progressHandle, fileLen, isWriteFile, downloadType)
	self:updatePath(origPath);
	self:addEventHandle(evtHandle, progressHandle);
end

function M:unload()
	if (nil ~= self.mEvtHandle) then
		self.mEvtHandle:clearEventHandle();
		self.mEvtHandle = nil;
	end

	if (self.mProgressEventDispatch ~= nil) then
		self.mProgressEventDispatch:clearEventHandle();
		self.mProgressEventDispatch = nil;
	end

	self:reset();
end

function M:onProgressEventHandle(dispObj)
	if(nil ~= self.mProgressEventDispatch) then
		self.mProgressEventDispatch:dispatchEvent(dispObj);
	end
end

return M;